File: Problem3_ContainerWater.java
// Time Complexity : o(n)
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution3 {
    public int maxArea(int[] height) {
        
        int low=0;
        int high=height.length-1;
        int max=0;
        while(low<high)
        {
            int lh=height[low];
            int rh=height[high];
            
            int width=high-low;
            int min_height=Math.min(lh,rh);
             max=Math.max(max,min_height*width);
            
            if(lh<rh)
                low++;
            else
                high--;
        }
        return max;
    }
}

File: Problem3_containerwater.py
'''
We will use a two pointer approach to do this. For getting a better area (area = height * wisth ) we know that widh will always decrease by 1 through consecutive buckets thus the deciding factor in this problem to get biggest container will be the height.
'''

'''
time complexity : O(n)
Space complexity : O(1)
'''


class Solution:
    def maxArea(self, height: List[int]) -> int:
        # handling the null case
        if (height == None or len(height) == 0):
            return 0
        n = len(height)
        # largest container
        maximum = 0
        low = 0
        high = n - 1

        # the min term in the code is the height and the difference term is the width
        while (low < high):
            maximum = max(maximum, min(height[low], height[high]) * (high - low))
            if (height[low] < height[high]):
                low += 1
            else:
                high -= 1
        return maximum